nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -234,52 +240,295 @@ impl Builtin for StringPrototypeIterator {
+        // 2. Let S be ? ToString(O).
+        let s = to_string(agent, o)?;
+        // 3. Let position be ? ToIntegerOrInfinity(pos).
+        let position = to_integer_or_infinity(agent, args.get(0))?.into_i64(agent);
+        // 4. Let size be the length of S.
+        let size = s.utf16_len(agent);
+        // 5. If position < 0 or position ≥ size, return undefined.
+        if position < 0 || position >= i64::try_from(size).unwrap() {
+            Ok(Value::Undefined)
+        } else {
+            // 6. Let cp be CodePointAt(S, position).
+            // 7. Return 𝔽(cp.[[CodePoint]]).
+            // TODO: Deal with lone surrogates.
+            let u8_idx = s
+                .utf8_index(agent, usize::try_from(position).unwrap())
+                .unwrap();
+            let ch = s.as_str(agent)[u8_idx..].chars().next().unwrap();
+            Ok(Value::from(ch as u32))
+        }
     }
 
-    fn get_timezone_offset(
-        _agent: &mut Agent,
-        _this_value: Value,
-        _: ArgumentsList,
-    ) -> JsResult<Value> {
-        todo!()
+    fn concat(agent: &mut Agent, this_value: Value, args: ArgumentsList) -> JsResult<Value> {
+        // 1. Let O be ? RequireObjectCoercible(this value).
+        let o = require_object_coercible(agent, this_value)?;
+        // 2. Let S be ? ToString(O).
+        let s = to_string(agent, o)?;
+        // 3. Let R be S.
+        let mut strings = Vec::with_capacity(args.len() + 1);
+        strings.push(s);
+        // 4. For each element next of args, do
+        //     a. Let nextString be ? ToString(next).
+        //     b. Set R to the string-concatenation of R and nextString.
+        for next in args.iter() {
+            strings.push(to_string(agent, *next)?);
+        }
+        // 5. Return R.
+        Ok(String::concat(agent, &strings).into_value())
+    }
+
+    fn ends_with(agent: &mut Agent, this_value: Value, args: ArgumentsList) -> JsResult<Value> {
+        // 1. Let O be ? RequireObjectCoercible(this value).
+        let o = require_object_coercible(agent, this_value)?;
+        // 2. Let S be ? ToString(O).
+        let s = to_string(agent, o)?;
+
+        // 3. Let isRegExp be ? IsRegExp(searchString).
+        // 4. If isRegExp is true, throw a TypeError exception.
+        // TODO
+
+        // 5. Let searchStr be ? ToString(searchString).
+        let search_str = to_string(agent, args.get(0))?;
+
+        // 6. Let len be the length of S.
+        // 7. If endPosition is undefined, let pos be len; else let pos be ? ToIntegerOrInfinity(endPosition).
+        // 8. Let end be the result of clamping pos between 0 and len.
+        let end_position = args.get(1);
+        let haystack_str = if end_position.is_undefined() {
+            s.as_str(agent)
+        } else {
+            let pos = to_integer_or_infinity(agent, end_position)?.into_usize(agent);
+            let end = if pos != 0 {
+                pos.min(s.utf16_len(agent))
```

Comment:
**issue:** Need to also clamp with 0 at the low end, in case `end_position` is negative.

Reply:
`pos` here is a usize, and `Number::into_usize` clamps negative numbers to 0. I'll add comments to make that clear.

Reply:
Alright, sorry for the noise.